docs/design/1161-standardize-stdlib.md

```diff
@@ -0,0 +1,197 @@
+# Proposal: Standardizing the Alloy standard library
+
+* Author: Robert Fratto (@rfratto)
+* Last updated: 2024-06-27
+* Discussion link: <https://github.com/grafana/alloy/pull/1161>
+
+## Abstract
+
+This proposal introduces a way to standardise on how new capabilities are added
+to Alloy's standard library so that usage of the standard library feels
+consistent.
+
+## Problem
+
+As of Alloy v1.2.0, there are 19 identifiers in the standard library:
+
+* `base64_decode`
+* `coalesce`
+* `concat`
+* `constants`
+* `env`
+* `format`
+* `join`
+* `json_decode`
+* `json_path`
+* `nonsensitive`
+* `replace`
+* `split`
+* `to_lower`
+* `to_upper`
+* `trim`
+* `trim_prefix`
+* `trim_space`
+* `trim_suffix`
+* `yaml_decode`
+
+These identifiers were added organically over time; there has been little
+thought to consistency across all identifiers.
+
+While 19 identifiers is manageable and possible for a user to remember the most
+important ones, if we don't establish a way to standardise existing and new
+identifiers, the inconsistency between identifiers will only grow over time as
+new identifiers are added and become harder to understand.
+
+This proposal aims to:
+
+* Establish a convention for how to name identifiers in the standard library.
+* Propose a way to align existing identifiers with the new convention without
+  breaking backwards compatibility.
+
+For the sake of demonstrating the convention, new identifiers are provided as
+examples, but are out of scope of this proposal and should not be considered
+accepted if this proposal is accepted.
+
+## Proposal
+
+The standard library will introduce namespaces for identifiers that group
+similar identifiers together:
+
+* Array functions
+    * `array.concat` (previously `concat`)
+* Conversion functions
+    * `convert.nonsensitive` (previously `nonsensitive`)
+* Decoding functions
+    * `decode.base64` (previously `decode_base64`)
+    * `decode.json` (previously `decode_json`)
+    * `decode.yaml` (previously `decode_yaml`)
+* OS-related functions
+    * `os.env` (previously `env`)
+* String functions
+    * `string.format` (previously `format`)
+    * `string.join` (previously `join`)
+    * `string.json_path` (previously `json_path`)
```

Comment:
This one feels a bit odd since it only works on a subtype of strings.

Would it make sense to have format namespaces. IE
`json.decode`, `yaml.decode`, and `json.path`?

Reply:
I agree that this one doesn't quite fit in. 

I suspect that it would be ideal if the namespaces we identify could _potentially_ have a handful of identifiers, otherwise we risk ending up with way too many namespaces that become hard to reason about. 

If we had formatting namespaces, I don't suspect we'd usually have more than 2-3 identifiers:

* `json.decode`
* `json.encode`
* `json.path` 
* `yaml.decode`
* `yaml.encode` 
* `base64.encode`
* `base64.decode`
* (etc) 

I'm concerned that these are really small namespaces and if we continue to add support for new encoding/decoding targets (MD5? SHA?), the number of top-level format namespaces would quickly outnumber everything else. 

But still, I agree this one sticks out as strange and it may be worth exploring all the different ways we could group together encoding-related functionality.

Reply:
I think that's totally reasonable and wouldn't fight for the `yaml`, `json` etc namespaces. Maybe a filter/search namespace? Do we fill that would have more options?

Reply:
Overall in favor of the proposed solution.

Reply:
As a temporary solution, we could leave `json_path` alone (like how I'm handling `constants` and `coalesce`) while we try to identify what the best namespace for it is

Reply:
I've updated the proposal so that `json_path` isn't given a namespace yet so we have more time to decide. I'm a little concerned that if we were to put `json_path`, `constants`, and `coalesce` into a namespace it would change the design of the other proposed namespaces, but I'm not sure how realistic that concern is.